### PR TITLE
[GlobalIsel][AArch64] MVP for vectorized selects.

### PR DESCRIPTION
**Pull request by tschuett**
Try to select Selects where the condition is a vector.

Inspired by arm64-vselect.ll and extensions to it.

Note that there are not tests for all legal types.

### CODE REVIEW

**Review by davemgreen — COMMENTED** [review]
I think for vectors we should be expanding the select into or+and+not during legalization, then selecting them with the existing bsp patterns. I think that's how it currently works, is it not working well in practice?

**Comment by tschuett** [comment]
I thought that legalizing them and manually selecting them generates better code. They are kept at higher abstraction in the pass pipeline. E.g. the generic combiner sees them as selects and not as binary operations.

File: llvm/test/CodeGen/AArch64/GlobalISel/legalize-select.mir [hunk 1]
```diff
@@ -53,13 +53,8 @@ body:             |
...
     ; CHECK-NEXT: [[ICMP:%[0-9]+]]:_(<2 x s32>) = G_ICMP intpred(sgt), [[COPY]](<2 x s32>), [[BUILD_VECTOR]]
     ; CHECK-NEXT: [[SEXT_INREG:%[0-9]+]]:_(<2 x s32>) = G_SEXT_INREG [[ICMP]], 1
-    ; CHECK-NEXT: [[C1:%[0-9]+]]:_(s32) = G_CONSTANT i32 -1
-    ; CHECK-NEXT: [[BUILD_VECTOR1:%[0-9]+]]:_(<2 x s32>) = G_BUILD_VECTOR [[C1]](s32), [[C1]](s32)
-    ; CHECK-NEXT: [[XOR:%[0-9]+]]:_(<2 x s32>) = G_XOR [[SEXT_INREG]], [[BUILD_VECTOR1]]
-    ; CHECK-NEXT: [[AND:%[0-9]+]]:_(<2 x s32>) = G_AND [[COPY1]], [[SEXT_INREG]]
-    ; CHECK-NEXT: [[AND1:%[0-9]+]]:_(<2 x s32>) = G_AND [[COPY]], [[XOR]]
-    ; CHECK-NEXT: [[OR:%[0-9]+]]:_(<2 x s32>) = G_OR [[AND]], [[AND1]]
-    ; CHECK-NEXT: $d0 = COPY [[OR]](<2 x s32>)
+    ; CHECK-NEXT: [[SELECT:%[0-9]+]]:_(<2 x s32>) = G_SELECT [[SEXT_INREG]](<2 x s32>), [[COPY1]], [[COPY]]
```
Comment (tschuett) [inline llvm/test/CodeGen/AArch64/GlobalISel/legalize-select.mir:56]: The combiner now sees a select and not the binary operators and a build vector.

**Comment by tschuett** [comment]
RegbankSelect has an incorrect optimization for vectorized selects that do not exist today.
https://github.com/llvm/llvm-project/blob/b203d5320df7754bf0ce019f01347a0ef743a207/llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp#L918

**Comment by tschuett** [comment]
The select to floating point minimax fold can never hit for vectorized selects. They do not exist today.
https://github.com/llvm/llvm-project/blob/5841140e38c94f3d2b29354f67e70803d88ce174/llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp#L6178

**Pull request closed without merging**